### PR TITLE
Localize menu media

### PR DESCRIPTION
**Pull request by wpride**

### CODE REVIEW

**Comment by wpride** [comment]
cross with https://github.com/dimagi/commcare-odk/pull/238

@ctsims

File: backend/src/org/commcare/suite/model/Callout.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.Hashtable;
+import java.util.Vector;
+
+/**
+ * Created by wpride1 on 4/14/15.
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/Callout.java]: class description/comment, please

File: backend/src/org/commcare/suite/model/Callout.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    */
+    public Callout evaluate(EvaluationContext context) {
+        return this;
+    }
+
+    /*
+     * (non-Javadoc)
+     * @see org.javarosa.core.util.externalizable.Externalizable#readExternal(java.io.DataInputStream, org.javarosa.core.util.externalizable.PrototypeFactory)
+     */
+    public void readExternal(DataInputStream in, PrototypeFactory pf) throws IOException, DeserializationException {
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/Callout.java]: don't need to save displayName?

File: backend/src/org/commcare/suite/model/Detail.java [hunk 3]
```diff
@@ -64,6 +65,22 @@ public Detail(
             String id, DisplayUnit title,
             Vector<Detail> details,
             Vector<DetailField> fields,
+            OrderedHashtable<String, String> variables, Action action, Callout callout
+    ) {
+        this(
+                id, title,
+                ArrayUtilities.copyIntoArray(details, new Detail[details.size()]),
+                ArrayUtilities.copyIntoArray(fields, new DetailField[fields.size()]),
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/Detail.java]: Why the need to call copyIntoArray? Can't the vectors be passed directly into the constructor defined below this one?

File: backend/src/org/commcare/suite/model/graph/DisplayData.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package org.commcare.suite.model.graph;
+
+/**
+ * Created by wpride1 on 4/24/15.
+ */
+public class DisplayData {
+
+    String name;
+    String iURI;
+    String aURI;
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/graph/DisplayData.java]: can these have full names (imageURI/audioURI) please?

File: backend/src/org/commcare/suite/model/graph/DisplayData.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package org.commcare.suite.model.graph;
+
+/**
+ * Created by wpride1 on 4/24/15.
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/graph/DisplayData.java]: class description/comment would be helpful

File: backend/src/org/commcare/xml/CalloutParser.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import org.commcare.suite.model.Callout;
+import org.javarosa.xml.ElementParser;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.kxml2.io.KXmlParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+
+/**
+ * Parser used in DetailParser to parse the Grid attributes for a GridEntityView
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/CalloutParser.java]: This seems not true.

File: backend/src/org/commcare/suite/model/Displayable.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package org.commcare.suite.model;
+
+/**
+ * Created by wpride1 on 4/27/15.
```
Comment (orangejenny) [inline backend/src/org/commcare/suite/model/Displayable.java]: class description please

File: backend/src/org/commcare/xml/CalloutParser.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+        String actionName = parser.getAttributeValue(null, "action");
+        String image = parser.getAttributeValue(null, "image");
+        String dislayName = parser.getAttributeValue(null, "name");
+
+        Callout callout = new Callout(actionName, image, dislayName);
+
+        while(nextTagInBlock("lookup")){
+            String tagname = parser.getName();
+            if(tagname != null && tagname.equals("extra")){
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/CalloutParser.java]: Minor, but `"extra".equals(tagname)` would make this more concise. Java and its endless checking for nulls...

File: backend/src/org/commcare/xml/CalloutParser.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    public Callout parse() throws InvalidStructureException, IOException, XmlPullParserException {
+
+        String actionName = parser.getAttributeValue(null, "action");
+        String image = parser.getAttributeValue(null, "image");
+        String dislayName = parser.getAttributeValue(null, "name");
+
+        Callout callout = new Callout(actionName, image, dislayName);
+
+        while(nextTagInBlock("lookup")){
+            String tagname = parser.getName();
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/CalloutParser.java]: conventions: tagname => tagName

File: backend/src/org/commcare/xml/CalloutParser.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    public CalloutParser(KXmlParser parser) {
+        super(parser);
+    }
+
+    public Callout parse() throws InvalidStructureException, IOException, XmlPullParserException {
+
+        String actionName = parser.getAttributeValue(null, "action");
+        String image = parser.getAttributeValue(null, "image");
+        String dislayName = parser.getAttributeValue(null, "name");
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/CalloutParser.java]: dislayName => displayName

File: backend/src/org/commcare/xml/CommCareElementParser.java [hunk 11]
```diff
@@ -46,18 +46,28 @@ public CommCareElementParser(KXmlParser parser) {
...
 
         while (nextTagInBlock("display")) {
             if (parser.getName().equals("text")) {
-                displayText = new TextParser(parser).parse();
+
+                String attributeValue = parser.getAttributeValue(null,"form");
+                
+                if(attributeValue != null && attributeValue.equals("image")){
+                    imageValue = new TextParser(parser).parse();
+                } else if(attributeValue != null && attributeValue.equals("audio")){
```
Comment (orangejenny) [inline backend/src/org/commcare/xml/CommCareElementParser.java]: Can you update the display element spec at https://github.com/dimagi/commcare/wiki/Suite20 ? It's hard to tell if the parser is doing the right thing here without being certain what kind of XML it's getting as input.

**Comment by orangejenny** [comment]
Minor comments.

General question: the commit list here was surprising - what do callouts have to do with localizing menus? Is this the right collection of commits?

**Comment by wpride** [comment]
@orangejenny yes I mixed these two up. I just fully merged them because I'd made the changes already on this branch. @phillipm changes from https://github.com/dimagi/commcare/pull/69#discussion_r29343752 made here. I know this is terrible but it'd take forever for me to separate them now

**Comment by phillipm** [comment]
@wpride updates for PR https://github.com/dimagi/commcare/pull/69 look good.
@orangejenny all of his commits for the callout feature are also in this PR, so when you are good to go, please also merge the cross-referenced https://github.com/dimagi/commcare-odk/pull/234

**Pull request merged**